### ISSUE TITLE
clang does not add executable rights to an existing output file on macOS

### ISSUE THREAD

**Issue by ThomasMader** [NixOS/nixpkgs#57120]
## Issue description

This is the continuation for the problem in https://github.com/NixOS/nixpkgs/pull/55998#issuecomment-465327570.
cc @veprbl 

The problem is that clang and gcc on macOS doesn't add executable rights if the output file was already created without executable rights.

### Steps to reproduce

cd /tmp
touch executableTest
echo "int main() {}" > executableTest.c
nix-shell -p clang --pure --run "clang executableTest.c -o executableTest"
./executableTest

It works with gcc but still not with clang if I checkout nixpkgs master at 84be644f87493ab5bc874580c1d45d7635d6547f go into the checkout directory and do the following:

cd /tmp
touch executableTest
echo "int main() {}" > executableTest.c
nix-shell -p gcc --pure --run "gcc executableTest.c -o executableTest"
./executableTest

## Technical details

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the
results.

 - system: `"x86_64-darwin"`
 - host os: `Darwin 18.2.0, macOS 10.14.3`
 - multi-user?: `no`
 - sandbox: `no`
 - version: `nix-env (Nix) 2.1.3`
 - channels(thomad): `"nixpkgs-19.03pre171497.cc1fa81489e"`
 - nixpkgs: `/Users/thomad/.nix-defexpr/channels/nixpkgs`

**Comment by veprbl** [comment]
And how does this tie into a problem with `dmd -run`? The D compiler calls clang?

**Comment by veprbl** [comment]
I tested gcc and I also don't get executable bit.

**Comment by ThomasMader** [comment]
Yes this is actually a linker error because dmd uses the system linker for linking only.

**Comment by veprbl** [comment]
I've updated the "Steps to reproduce" because "nix-shell '<nixpkgs>' -A dmd --pure --run "gcc executableTest.c -o executableTest" fails with "gcc: command not found"

**Comment by veprbl** [comment]
I'm not sure if linker is obliged to set any file permissions. I think I remember this was not always the case that this would be done. Perhaps dmd should be responsible for setting correct permissions on files it wants to execute.

**Comment by ThomasMader** [comment]
That's the case on every system out there it seems, even on NixOS (Linux) itself it was never a problem and currently also works.

**Comment by ThomasMader** [comment]
Just to show that it is indeed the linker. Decoupling the object creation from the linker step.

```
clang -c executableTest.c
ld -o executableTest executableTest.o -lSystem
```

Installing clang with `nix-env -i clang` and trying `clang -v executableTest.c -o executableTest` works because in this case clang uses the system linker of macOS.
This is strange as it is not a pure environment.

**Comment by veprbl** [comment]
Looking at the source code of cctools-port I see that the declared behaviour seems to be to set permissions even on existing files:
https://github.com/tpoechtrager/cctools-port/blob/f53c6186393915ba9bff287f18769f9e14ec02e6/cctools/ld64/src/ld/OutputFile.cpp#L2769-L2778

**Comment by ThomasMader** [comment]
The problem is that this code has a filesystem check for hfs which fails on my macOS machine because it's the new apfs fs.
Because of this outputIsMappableFile is false and the code uses open were the permissions are not changed if the file already existed. (https://github.com/tpoechtrager/cctools-port/blob/f53c6186393915ba9bff287f18769f9e14ec02e6/cctools/ld64/src/ld/OutputFile.cpp#L2864)

I wrote the following code to verify the fs type problem.

```
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <sys/param.h>
#include <sys/mount.h>
#include <stdio.h>

int main()
{
        struct stat stat_buf;
        if (stat("/tmp/executableTest", &stat_buf) != -1) {
                if (stat_buf.st_mode & S_IFREG) {
                        printf("Normal file\n");
                        struct statfs fsInfo;
                        if ( statfs("/tmp/executableTest", &fsInfo) != -1 ) {
                                if ( strcmp(fsInfo.f_fstypename, "hfs") == 0) {
                                        printf("Output is mappable file\n");
                                } else {
                                        printf("Not hfs fs: %s\n", fsInfo.f_fstypename);
                                }
                        } else {
                                printf("Output is not a mappable file\n");
                        }
                } else {
                        printf("Not a normal file\n");
                }
        } else {
                printf("Can not stat file\n");
        }
}
```
The system linker clearly must have different code.
Nix really needs to switch to the system packages otherwise those kind of problems will continue to popup.

**Comment by veprbl** [comment]
So cctools-port that we use is based on ld64 274.2, apple has released 409.12 [1] which seems to include apfs support just as you said:
```c++
if ( (strcmp(fsInfo.f_fstypename, "hfs") == 0) || (strcmp(fsInfo.f_fstypename, "apfs") == 0) ) {
```

We would need to bump that in cctools-port upstream.

[1] https://opensource.apple.com/source/ld64/ld64-409.12/

**Comment by veprbl** [comment]
cc @matthewbauer

**Comment by matthewbauer** [comment]
Yeah it makes sense to bump ld64 for this (there are lots of other reasons to want the newer stuff too). I looked at it a while ago but never finished it (now noticing there are a few places I merged badly)

https://github.com/matthewbauer/cctools-port/commits/cctools-921-ld64-409.12

If anyone wants to take that up feel free. Otherwise, we might be able to get a quick patch that adds apfs support to ld64.

**Comment by ThomasMader** [comment]
@matthewbauer I want to bring up the issue that Nix doesn't use the official SDK from Apple again.

In https://github.com/NixOS/nixpkgs/issues/54586#issuecomment-457731207 you mentioned that the reasoning behind that is that you don't want users to have to accept the XCode agreement.
I want to list some arguments against that.

1) I am not a lawyer but to me it seems that by downloading any part of the official SDK I already fall into the license agreement. It's stated right at the beginning of https://www.apple.com/legal/sla/docs/xcode.pdf -> _OR BY
DOWNLOADING, USING OR COPYING ANY PART OF THIS APPLE SOFTWARE_
If you are patching up cctools-port or any headers distributed via Nix by looking at the code of the official SDK you are already introducing the need for all users to agree to the license.

2) It's not possible to create a system which is 100% compatible with the official SDK without looking at their sources. If you are not compatible, all the software you are distributing will have problems. This issue and the problem in https://github.com/NixOS/nixpkgs/issues/54586#issuecomment-457731207 are for sure not even the tip of the iceberg.

If you ever want to have a production ready Nix distribution on macOS I see no other way than to go with the official sources.
I only maintain dlang packages here on Nix but I really don't want to lose valuable time by figuring out those kind of bugs. It's a waste of time and I fail to see any reasoning for the status quo.

**Comment by matthewbauer** [comment]
cctools is released by Apple as free software so we don't think there is any issue there. Our usage of some SDKs like Cocoa and Foundation is a legitimate concern though.

There are other benefits besides the licensing issue though. Because we build the toolchain ourselves, we aren't tied to the often out-of-date toolchain that is preinstalled on macOS. The user doesn't even need CommandLineTools or Xcode.app installed, because the whole toolchain is free software and they can just download it from the binary cache. We don't have to rely on the user having the right Xcode version installed before we can compile things.

I definitely want Nix on macOS to be production ready. For things like C/C++ and Haskell I would hesitantly say it is but other languages there is definitely a lot of work to do. I definitely understand the frustration with the differences from the official Apple SDK though.

In this case, I'm not sure if using the official SDK would completely fix the issue. If you had a pre-APFS Xcode version installed and updated to High Sierra you would see this problem too.

**Comment by veprbl** [comment]
A fix for this was added to #56744

**Comment by ThomasMader** [comment]
> cctools is released by Apple as free software so we don't think there is any issue there. Our usage of some SDKs like Cocoa and Foundation is a legitimate concern though.

Why are you using cctools-port then?

> There are other benefits besides the licensing issue though. Because we build the toolchain ourselves, we aren't tied to the often out-of-date toolchain that is preinstalled on macOS. The user doesn't even need CommandLineTools or Xcode.app installed, because the whole toolchain is free software and they can just download it from the binary cache. We don't have to rely on the user having the right Xcode version installed before we can compile things.

Yes that's a nice feature but that doesn't go away if you just bundle the official SDK. The latest version of the SDK should be bundled. Supporting the latest versions seems much more important than older ones.
Bundling multiple versions of the SDK results in a lot of support complexity for nixpkgs on Darwin.
According to https://en.wikipedia.org/wiki/Xcode#10.x_series with 10.1 you would support High Sierra and up. Thats good. If one doesn't want to update he also needs to use an older version of nixpkgs.

> I definitely want Nix on macOS to be production ready. For things like C/C++ and Haskell I would hesitantly say it is but other languages there is definitely a lot of work to do. I definitely understand the frustration with the differences from the official Apple SDK though.

How can it ever be production ready on C/C++ if you bundle development libraries in a version not used by Apple? You are creating your own macOS development environment which might be considered production ready but because upstream projects develop and test against the official SDK you are opening the door for chaos.
Even if you get all packages to build there will be a lot of runtime issues hidden because no one supports you at testing.

> In this case, I'm not sure if using the official SDK would completely fix the issue. If you had a pre-APFS Xcode version installed and updated to High Sierra you would see this problem too.

If I update the OS I need to expect that things might not work anymore and I need to update development libraries too. If my project still works I am lucky but there might be bugs no one knows about yet. I might even need to fix my project, that's how it is.

**Comment by ThomasMader** [comment]
> Installing clang with nix-env -i clang and trying clang -v executableTest.c -o executableTest works because in this case clang uses the system linker of macOS.
This is strange as it is not a pure environment.

That's probably also an issue right? Can somebody reproduce this problem?

**Comment by veprbl** [comment]
> That's probably also an issue right? Can somebody reproduce this problem?

Can you check priority of ~/.nix-profile/bin in your $PATH?

**Comment by ThomasMader** [comment]
1) Installing clang takes a couple of minutes on my machine. It's a Mac Mini with HDD and just 4GB so I am used to it but that was really too long.
The process needed nearly 1GB of memory before it terminated with the following output.

```
nix-env -i clang
warning: there are multiple derivations named 'clang-5.0.2'; using the first one
installing 'clang-5.0.2'
```
2) Couple of outputs:

```
$ which clang
/Users/thomad/.nix-profile/bin/clang
$ clang -v test.c -o test
clang version 5.0.2 (tags/RELEASE_502/final)
Target: x86_64-apple-darwin18.2.0
Thread model: posix
InstalledDir: /Users/thomad/.nix-profile/bin
 "/nix/store/ap8srzgnkxcbq0zmsxkiazd7z1zaljb4-clang-5.0.2/bin/clang-5.0" -cc1 -triple x86_64-apple-macosx10.14.0 -Wdeprecated-objc-isa-usage -Werror=deprecated-objc-isa-usage -emit-obj -mrelax-all -disable-free -disable-llvm-verifier -discard-value-names -main-file-name test.c -mrelocation-model pic -pic-level 2 -mthread-model posix -mdisable-fp-elim -masm-verbose -munwind-tables -target-cpu penryn -target-linker-version 274.2 -v -dwarf-column-info -debugger-tuning=lldb -nostdsysteminc -resource-dir /nix/store/ap8srzgnkxcbq0zmsxkiazd7z1zaljb4-clang-5.0.2/lib/clang/5.0.2 -fdebug-compilation-dir /tmp -ferror-limit 19 -fmessage-length 80 -stack-protector 1 -fblocks -fobjc-runtime=macosx-10.14.0 -fencode-extended-block-signature -fmax-type-align=16 -fdiagnostics-show-option -fcolor-diagnostics -o /var/folders/wq/m1dnr42s42n5msqk2v8l0lfc0000gn/T/test-4d0a98.o -x c test.c
clang -cc1 version 5.0.2 based upon LLVM 5.0.2 default target x86_64-apple-darwin18.2.0
#include "..." search starts here:
#include <...> search starts here:
 /nix/store/ap8srzgnkxcbq0zmsxkiazd7z1zaljb4-clang-5.0.2/lib/clang/5.0.2/include
End of search list.
 "/usr/bin/ld" -demangle -lto_library /nix/store/ap8srzgnkxcbq0zmsxkiazd7z1zaljb4-clang-5.0.2/lib/libLTO.dylib -no_deduplicate -dynamic -arch x86_64 -macosx_version_min 10.14.0 -o test /var/folders/wq/m1dnr42s42n5msqk2v8l0lfc0000gn/T/test-4d0a98.o -lSystem
$ echo $PATH 
/Users/thomad/.nix-profile/bin:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin
```

**Comment by veprbl** [comment]
> `nix-env -i clang`

Consider using `-iA`. The plain `-i` evaluates all derivations to lookup by derivation name this will be slower. The `-i`-like functionality is not provided by the new `nix` command line interface for that and other reasons.

The second reason why you shouldn't use `-i` to install a compiler has to do with how we do compilation in nixpkgs. The package with the name "clang" is a mostly vanilla compiler (we call it unwrapped), it is accessible via `clang.cc` attribute. When you use stdenv to do an actual build we don't want to use a standard compiler, we use wrappers that modify command line options to enforce some of the sandboxing as well as adding additional common flags (e.g. flags to use the right toolchain). That wrapper is accessible via `clang` attribute and it has name "clang-wrapper". If you want a "pure" compiler you should be looking to install the wrapper, which works nicely outside the nixpkgs build and picks up the correct linker.

Perhaps it would make sense to rename "clang" to "clang-unwrapped" and keep "clang-wrapped" as "clang-wrapped".

**Comment by ThomasMader** [comment]
The fixes of #56744 are now merged into master and this issue is fixed.
